Fix #464 v4, story table row matches on wrong trigger var when there's

plocket opened the pull request:
[Just like #467, but for v4 branch. I'm not sure I'm loving this workflow, but hopefully we'll be able to drop v3 soon.]

only one row for the relevant variable. Closes #464.

E.g. When, in the story table, there should be
```
    | x[i].name.first | Sam | proxy_list[0].name.first |
    | x[i].name.first | Pat | proxy_list[1].name.first |
```

there's only
```
    | x[i].name.first | Sam | proxy_list[0].name.first |
```

the test shouldn't match on the page that's setting `proxy_list[1].name.first`, but it does, and it fills in "Sam" for the value.

Also
- enhance reports about the problem.
- avoids trigger var check entirely when it's not from a story table.

Edit:
Made some report strings more friendly, but they only get triggered in our unit tests
(which are faster and more focused than our integrated tests) and our unit tests don't
log or save their reports, so I'm providing sample output here of the three fancy reports:

```
Do you have a duplicate story table row? Multiple story table rows match this field's variable name. The test will use the last row that matched. See https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/automated_integrated_testing/#trigger.
The matching rows:
      | text_input | Some one-line text |  |
      | text_input | Some conflicting text |  |
This field's data:
{"selector":"#daquestion input[name=\"dGV4dF9pbnB1dA\"][id=\"dGV4dF9pbnB1dA\"][class=\"form-control\"]","tag":"input","guesses":[{"var":"text_input","value":""}],"type":"text","trigger":"direct_standard_fields"}
```
```
Is this story table missing a row? 1 story table row(s) matched a field name, but none of their trigger variable name(s) matched the trigger variable name in the screen's HTML. The test will have to ignore this field. See https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/automated_integrated_testing/#trigger.
The data of row(s) with conflicting trigger name(s):
      | x[i].name.first | Firstname | proxy_list[0].name.first |
This field's data:
{"selector":"#daquestion input[name=\"eFtpXS5uYW1lLmZpcnN0\"][id=\"eFtpXS5uYW1lLmZpcnN0\"][class=\"form-control\"]","tag":"input","guesses":[{"var":"x[i].name.first","value":""}],"type":"text","trigger":"proxy_list[1].name.first"}
```
```
This story table row needs a value in the "trigger" column, but it is empty. The test might get unexpected results. See https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/automated_integrated_testing/#trigger.
The row's data:
      | x[i].name.first | Firstname |  |
This field's data:
{"selector":"#daquestion input[name=\"eFtpXS5uYW1lLmZpcnN0\"][id=\"eFtpXS5uYW1lLmZpcnN0\"][class=\"form-control\"]","tag":"input","guesses":[{"var":"x[i].name.first","value":""}],"type":"text","trigger":"proxy_list[1].name.first"}
```

plocket committed:
Fix #464, story table row matches on wrong trigger var when there's

only one row for the relevant variable. Closes #464.

Also
- enhance reports about the problem.
- avoids trigger var check entirely when it's not from a story
table

plocket committed:
Merge branch 'releases/v4' into fix_trigger_match_v4

plocket commented on lib/scope.js:
```diff
@@ -826,36 +838,73 @@ module.exports = {
 
-    // If field uses proxies and there has been more than one match
-    // we now use the trigger var to narrow down the matches.
-    if ( uses_proxies && matches.length > 1 ) {
```
This was the line that was causing the problem, but fixing it meant I had to change other things so they wouldn't break.

plocket committed:
Merge branch 'releases/v4' into fix_trigger_match_v4

rpigneri-vol commented on lib/scope.js:
```diff
@@ -878,18 +885,23 @@ module.exports = {
 
         if ( row_matches ) {
           matches.push( var_datum );  // Collect all matches to start with.
+          // If the field uses proxy variables - x, i, j, etc, flag it. See more notes below.
```
We have decided to move this to above 892.

rpigneri-vol commented on lib/scope.js:
```diff
@@ -909,36 +921,73 @@ module.exports = {
+    // story tables need this disambiguation.
+    if ( trigger_might_be_needed && uses_proxies && matches.length > 0 ) {
+
+      if ( !field.trigger ) {
```
We have decided to change "field.trigger" to "field.triggerName" for clarity.

rpigneri-vol commented on lib/scope.js:
```diff
@@ -909,36 +921,73 @@ module.exports = {
-          // TODO: Add a warning if they're both `undefined`?
-        else {
-
+        if ( !match.trigger ) {
```
Same as above: we have decided to consider calling this match.triggerName.

plocket replied:
This may become a future issue to do properly because it goes deep. We may be able to do something more superficial in here, though.

rpigneri-vol commented on lib/scope.js:
```diff
@@ -878,18 +885,23 @@ module.exports = {
+          if ( var_datum.trigger === `ALKiln: no trigger variable needed` ) {
+            trigger_might_be_needed = false;
           }
           continue;
```
This line could be removed as it is not changing the flow of control.

rpigneri-vol commented on lib/scope.js:
```diff
@@ -909,36 +921,73 @@ module.exports = {
+            + `\nThe row's data is\n${ JSON.stringify( match.original )}.`
+            + `\nThis field's data is\n${ JSON.stringify( field )}`;
+          await scope.addToReport(scope, { type: `warning`, value: row_trigger_warning });
+          fallbacks.push( match );
```
Collecting fallbacks, but not yet using.

We might be able to provide the developer with the variable names and values that we are setting them to.

plocket replied:
Actually, I forgot the values given to the field will be visible in the info of the report later. Also, I think I misread the code. `\nThe row's data is\n${ JSON.stringify( match.original )}.` shows the data of that particular field. Resolving this for now, but let me know if we meant something different.

rpigneri-vol commented on lib/scope.js:
```diff
@@ -909,36 +921,73 @@ module.exports = {
-      if ( field.trigger === undefined ) { await scope.addToReport(scope, { type: `warning`, value: `A field on this page uses an index variable or a generic object, but the page does not contain an element with the trigger variable's name. Tell the developer to add an element to the page with the id "trigger"` }); }
+    // Make sure the trigger var matches for a page that uses proxies. Only
+    // story tables need this disambiguation.
+    if ( trigger_might_be_needed && uses_proxies && matches.length > 0 ) {
```
We are thinking that it might be easier to rename "matches" to "initial_matches" to avoid changing the meaning of the matches variable halfway through.

rpigneri-vol commented on lib/scope.js:
```diff
@@ -909,36 +921,73 @@ module.exports = {
+
+      // Feedback for devs on multiple row match without any matching triggers
+      if ( original_matches.length > 1 && matches.length === 0 ) {
+        let missing_row_warning = `The story table has multiple rows `
```
We have decided to put "Is the story table missing a row?" as the first sentence.

rpigneri-vol commented on lib/scope.js:
```diff
@@ -909,36 +921,73 @@ module.exports = {
+    // Dev may have accidentally included a row more than once.
     if ( matches.length > 1 ) {
-      await scope.addToReport(scope, { type: `warning`, value: `there are multiple story data matches for this field. The story table data that appears last will be used. The field's data is\n${JSON.stringify( field )}` });
+      let multiple_matches_msg = `Multiple story table rows match this field's `
```
We have decided to add "Do you have a duplicate story table row?" to the front of this error message.

rpigneri-vol reviewed (approved):
I need additional DA context, but there is nothing too obvious here that is a cause for concern.

plocket committed:
Refactor changes for clarity as per comments. Also add more tags

to tests for testing convenience. Also modified a few test fixures so they could survive
the new and improved report logs. The report strings only get triggered in our unit tests
(which are faster and more focused than our integrated tests) and our unit tests don't
log their reports, so I'm proving sample output here of the three fancy reports:

Do you have a duplicate story table row? Multiple story table rows match this field's variable name. The test will use the last row that matched. See https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/automated_integrated_testing/#trigger.
The matching rows:
      | text_input | Some one-line text |  |
      | text_input | Some conflicting text |  |
This field's data:
{"selector":"#daquestion input[name=\"dGV4dF9pbnB1dA\"][id=\"dGV4dF9pbnB1dA\"][class=\"form-control\"]","tag":"input","guesses":[{"var":"text_input","value":""}],"type":"text","trigger":"direct_standard_fields"}

Is this story table missing a row? 1 story table row(s) matched a field name, but none of their trigger variable name(s) matched the trigger variable name in the screen's HTML. The test will have to ignore this field. See https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/automated_integrated_testing/#trigger.
The data of row(s) with conflicting trigger name(s):
      | x[i].name.first | Firstname | proxy_list[0].name.first |
This field's data:
{"selector":"#daquestion input[name=\"eFtpXS5uYW1lLmZpcnN0\"][id=\"eFtpXS5uYW1lLmZpcnN0\"][class=\"form-control\"]","tag":"input","guesses":[{"var":"x[i].name.first","value":""}],"type":"text","trigger":"proxy_list[1].name.first"}

This story table row needs a value in the "trigger" column, but it is empty. The test might get unexpected results. See https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/automated_integrated_testing/#trigger.
The row's data:
      | x[i].name.first | Firstname |  |
This field's data:
{"selector":"#daquestion input[name=\"eFtpXS5uYW1lLmZpcnN0\"][id=\"eFtpXS5uYW1lLmZpcnN0\"][class=\"form-control\"]","tag":"input","guesses":[{"var":"x[i].name.first","value":""}],"type":"text","trigger":"proxy_list[1].name.first"}

plocket committed:
Merge branch 'releases/v4' of github.com:SuffolkLITLab/ALKiln into fix_trigger_match_v4

plocket committed:
Merge branch 'releases/v4' of github.com:SuffolkLITLab/ALKiln into fix_trigger_match_v4

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -65,6 +65,10 @@ module.exports = {
     let var_str = await scope.toTrimmedJSON( scope, { str: original.var });
     let value_str = await scope.toTrimmedJSON( scope, { str: original.value });
     let trigger_str = await scope.toTrimmedJSON( scope, { str: original.trigger });
+    if ( trigger_str === `ALKiln: no trigger variable needed` ) {
```
Suggestion: make this string a private variable elsewhere in the module. Right now, it's the same magic string being checked against multiple times, which makes it difficult if the string is changed in one place but not the other.

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -818,9 +822,14 @@ module.exports = {
   getMatchingRows: async function ( scope, { field, var_data }) {
+
+    `PLEASE READ THE COMMENTS IN THIS FUNCTION. It's complicated.`;
+
```
In my opinion, I'd remove this.

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -833,116 +842,183 @@ module.exports = {
+          // Check this until the value is set to `true` (if it ever is)
+          if ( !uses_proxies ) {
+            // If the field uses proxy variables - x, i, j, etc, flag it. See "proxies" note.
+            // AVOID `.test()` - check the MDN. You can have unexpected results.
```
```suggestion
```
Just remove this comment, shouldn't need to justify use of the standard library.

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -833,116 +842,183 @@ module.exports = {
         }
+      }  // ends for each a_row_that_matches_this_field
+
+      // Be very permissive, based on feedback. See #464 comment.
```
```suggestion
      // Be very permissive, based on feedback. See https://github.com/SuffolkLITLab/ALKiln/issues/464#issuecomment-1005166616
```

BryceStevenWilley commented:
I'll take a closer look at scope.js and the larger changes there soon!

plocket committed:
Merge branch 'releases/v4' into fix_trigger_match_v4

plocket committed:
Remove distracting comment

Co-authored-by: Bryce Willey <Bryce.Steven.Willey@gmail.com>

plocket committed:
Use specific comment link

Co-authored-by: Bryce Willey <Bryce.Steven.Willey@gmail.com>

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -841,116 +850,182 @@ module.exports = {
+        if ( this_row_matches_this_field ) {
+          rows_that_match_field_names.push( table_row );
+          // Check this until the value is set to `true` (if it ever is)
+          if ( !uses_proxies ) {
+            // If the field uses proxy variables - x, i, j, etc, flag it. See "proxies" note.
+            uses_proxies = guess.var.match( proxies_regex );
           }
```
Making this a one liner is a bit more clear; the if makes it seem like something depends on the initial value (which is always false). 
```suggestion
          // If the field uses proxy variables - x, i, j, etc, flag it. See "proxies" note.
          uses_proxies = uses_proxies || guess.var.match( proxies_regex );
```

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -841,116 +850,182 @@ module.exports = {
-          continue;
-        }
+
+          let table_row_trigger_name = table_row.trigger;
+          if ( table_row_trigger_name === `ALKiln: no trigger variable needed` ) {
```
IMO no need to define a variable with a longer name than the original variable that's only used once.
```suggestion
          if ( table_row.trigger === `ALKiln: no trigger variable needed` ) {
```

plocket replied:
This was actually part of an effort I forgot about as I was going - an intermediate step for improving `.trigger` to `.trigger_name`.

plocket replied:
Maybe I should save it for a future, more extensive, PR.

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -841,116 +850,182 @@ module.exports = {
+        await scope.addToReport(scope, { type: `warning`, value: page_trigger_warning });
+      }
+
+      for ( let a_row_that_matches_this_field of rows_that_match_field_names ) {
```
```suggestion
      for ( let row_that_matches_this_field of rows_that_match_field_names ) {
```
and elsewhere

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -841,116 +850,182 @@ module.exports = {
 
-    // There are mutliple guesses because field names are encoded
+    let rows_that_match_field_names = [];
+    let final_matching_rows = [];
```
This var isn't used for about 100 lines, I'd move it closer to the relevant code.

plocket replied:
I'll move it as far down as I can, though it has to be kept in the outermost scope. In js, you have to define things outside of `if` statements or they later aren't defined in the scope outside of the `if` brackets. [Which you may already know, just figured I'd include that in case it was useful.]

BryceStevenWilley reviewed (approved):
LGTM, some optional comments that I found going through it more slowly.

plocket committed:
Implement review comments:

* Remove spurious instructions
* Create state var to contain flag for non-story-table row
* Move var instantiation closer to use of var

plocket committed:
Clarify use of `use_proxies`

Co-authored-by: Bryce Willey <Bryce.Steven.Willey@gmail.com>

plocket committed:
Remove single use of intermediate variable

plocket committed:
Shorten var name

plocket committed:
Remove use of env_vars (probably got in through an old merge)

BryceStevenWilley commented on lib/steps.js:
```diff
@@ -66,6 +66,8 @@ let click_with = {
   pc: 'click',
 };
 
+scope.row_needs_trigger_flag = `ALKiln: no trigger variable needed`;
```
We may have mentioned this before, but what's the reason that this string isn't just defined inside `scope.js`? You still have the problem on needing to define it in each file that uses scope (steps and the unit test).

plocket replied:
It was partly because I thought it was being used in steps.js (which it wasn't) and partly because I was instantiating it in the script where other state variables were being defined. I haven't defined any `scope` props in scope.js other than functions, but I can switch it to there.

plocket replied:
Thanks for the second eye!

plocket committed:
Move state flag to state file
